library/std/src/sys_common/wtf8.rs

```diff
@@ -474,13 +474,12 @@ impl Wtf8Buf {
         Wtf8Buf { bytes: bytes.into_vec(), is_known_utf8: false }
     }
 
-    /// Part of a hack to make PathBuf::push/pop more efficient.
+    /// More well behaving alternative to allowing outer types
+    /// full mutable access to the core `Vec`.
+    /// Provides plumbing to core `Vec::extend_from_slice`.
     #[inline]
-    pub(crate) fn as_mut_vec_for_path_buf(&mut self) -> &mut Vec<u8> {
-        // FIXME: this function should not even exist, as it implies violating Wtf8Buf invariants
-        // For now, simply assume that is about to happen.
-        self.is_known_utf8 = false;
-        &mut self.bytes
+    pub(crate) fn extend_from_slice(&mut self, other: &[u8]) {
+        self.bytes.extend_from_slice(other);
```

Comment:
This will need to also do this:
```suggestion
    pub(crate) fn extend_from_slice(&mut self, other: &[u8]) {
        *self.is_known_utf8 = false;
        self.bytes.extend_from_slice(other);
```

Or rerun the validation? In that case it would look more like


```rust
    pub(crate) fn extend_from_slice(&mut self, other: &[u8]) {
        if *self.is_known_utf8 {
            *self.is_known_utf8 = validate_combined(self, other);
        }
        self.bytes.extend_from_slice(other);
```

Reply:
It seems as though there's no `validate_combined` function already in scope... Which, to me, seems a little silly. How would we feel about me adding it? :eyes:

Reply:
Go for it, it was just an example name.

Though if `self` is known to be UTF-8, validating `other` as UTF-8 would be sufficient.

Reply:
Sounds like a good way to top it all off! I'll get that done tomorrow.